### PR TITLE
Optimize hprod in sparse hessian

### PR DESCRIPTION
**Pull request by tmigot**
@amontoison This is not ideal for the objective Hessian, and I have done it quickly, but it should work :).

### CODE REVIEW

File: src/sparse_hessian.jl [hunk 1]
```diff
@@ -61,14 +92,29 @@ end
...
+  y::AbstractVector,
   vals::AbstractVector,
-)
+) where {Tag, GT, S, T}
   nvar = length(x)
+  ncon = length(y)
+
+  b.sol[1:ncon] .= zero(T) # cx
+  b.sol[(ncon + 1):(ncon + nvar)] .= x
+  b.sol[(ncon + nvar + 1):(2 * ncon + nvar)] .= y
```
Comment (amontoison) [inline src/sparse_hessian.jl:106]:
```suggestion
  view(b.sol, 1:ncon) .= zero(T) # cx
  view(b.sol, (ncon + 1):(ncon + nvar)) .= x
  view(b.sol, (ncon + nvar + 1):(2 * ncon + nvar)) .= y
```

**Review by amontoison — CHANGES_REQUESTED** [review]

**Comment by amontoison** [comment]
It is still slow compared to JuMP :(
```julia
julia> include("bench.jl")
Hessian of Lagrangian
ADNLP
  3.635 ms (4370 allocations: 8.58 MiB)
JuMP
  58.046 μs (137 allocations: 24.08 KiB)
Jacobian of constraints
ADNLP
  10.219 μs (22 allocations: 20.75 KiB)
JuMP
  14.944 μs (15 allocations: 19.80 KiB)
```

File: src/sparse_hessian.jl (hunk 1 above)
Comment (tmigot) [reply src/sparse_hessian.jl:106]: I have tested this, and it does not change the allocations.
```
using ADNLPModels, NLPModels; f(x) = sum(x); nlp = ADNLPModel(f, ones(2)); hess(nlp, nlp.meta.x0, nlp.meta.y0); @allocated hess(nlp, nlp.meta.x0, nlp.meta.y0)
```

File: src/sparse_hessian.jl [hunk 2]
```diff
@@ -61,14 +92,29 @@ end
...
+  b.sol[end] = obj_weight
+
+  b.longv .= 0
+
   for icol = 1:(b.ncolors)
     b.d .= (b.colors .== icol)
-    ForwardDiff.derivative!(b.res, t -> ForwardDiff.gradient(ℓ, x + t * b.d), 0)
+    b.longv[(ncon + 1):(ncon + nvar)] .= b.d
```
[the remark this answers is not in the document]
Comment (tmigot) [reply src/sparse_hessian.jl:113]: same as above

**Comment by tmigot** [comment]
> It is still slow compared to JuMP :(
> 
> ```julia
> julia> include("bench.jl")
> Hessian of Lagrangian
> ADNLP
>   3.635 ms (4370 allocations: 8.58 MiB)
> JuMP
>   58.046 μs (137 allocations: 24.08 KiB)
> Jacobian of constraints
> ADNLP
>   10.219 μs (22 allocations: 20.75 KiB)
> JuMP
>   14.944 μs (15 allocations: 19.80 KiB)
> ```

Can you check that you used the right branch for this test? Because, the allocation result is exactly the same as https://github.com/JuliaSmoothOptimizers/ADNLPModels.jl/issues/165#issuecomment-1599374463

Will https://github.com/JuliaSmoothOptimizers/ADNLPModels.jl/pull/166 bring any improvement?

File: src/sparse_hessian.jl (hunk 1 above)
Comment (dpo) [reply src/sparse_hessian.jl:106]: On the left-hand side, a view is always used I believe.

File: src/sparse_hessian.jl [hunk 3]
```diff
@@ -94,8 +141,9 @@ function hess_coord!(
   obj_weight::Real,
   vals::AbstractVector,
 )
-  ℓ = get_lag(nlp, b, obj_weight)
-  sparse_hess_coord!(ℓ, b, x, vals)
+  y = zeros(eltype(x), nlp.meta.nnln)
```
Comment (amontoison) [inline src/sparse_hessian.jl:144]: Is it possible to preallocate this vector?

**Review by amontoison — APPROVED** [review]

**Pull request merged**